adapters/smaato/smaato.go

```diff
@@ -214,75 +157,183 @@ func (a *SmaatoAdapter) MakeBids(internalRequest *openrtb2.BidRequest, externalR
+		return admType, nil
+	} else if strings.HasPrefix(adMarkup, `{"image":`) {
+		return smtAdTypeImg, nil
+	} else if strings.HasPrefix(adMarkup, `{"richmedia":`) {
+		return smtAdTypeRichmedia, nil
+	} else if strings.HasPrefix(adMarkup, `<?xml`) {
+		return smtAdTypeVideo, nil
+	} else {
+		return "", &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Invalid ad markup %s.", adMarkup),
+		}
+	}
+}
+
 func renderAdMarkup(adMarkupType adMarkupType, adMarkup string) (string, error) {
-	var markupError error
-	var adm string
 	switch adMarkupType {
 	case smtAdTypeImg:
-		adm, markupError = extractAdmImage(adMarkup)
+		return extractAdmImage(adMarkup)
 	case smtAdTypeRichmedia:
-		adm, markupError = extractAdmRichMedia(adMarkup)
+		return extractAdmRichMedia(adMarkup)
 	case smtAdTypeVideo:
-		adm, markupError = adMarkup, nil
+		return adMarkup, nil
 	default:
-		return "", fmt.Errorf("Unknown markup type %s", adMarkupType)
+		return "", &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unknown markup type %s.", adMarkupType),
+		}
 	}
-	return adm, markupError
 }
 
-func markupTypeToBidType(markupType adMarkupType) (openrtb_ext.BidType, error) {
-	switch markupType {
+func convertAdMarkupTypeToMediaType(adMarkupType adMarkupType) (openrtb_ext.BidType, error) {
+	switch adMarkupType {
 	case smtAdTypeImg:
 		return openrtb_ext.BidTypeBanner, nil
 	case smtAdTypeRichmedia:
 		return openrtb_ext.BidTypeBanner, nil
 	case smtAdTypeVideo:
 		return openrtb_ext.BidTypeVideo, nil
 	default:
-		return "", fmt.Errorf("Invalid markupType %s", markupType)
+		return "", &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unknown markup type %s.", adMarkupType),
+		}
 	}
 }
 
-func getAdMarkupType(response *adapters.ResponseData, adMarkup string) (adMarkupType, error) {
-	if admType := adMarkupType(response.Headers.Get("X-SMT-ADTYPE")); admType != "" {
-		return admType, nil
+func prepareCommonRequest(request *openrtb2.BidRequest) error {
+	if err := setUser(request); err != nil {
+		return err
 	}
-	if strings.HasPrefix(adMarkup, `{"image":`) {
-		return smtAdTypeImg, nil
+
+	if err := setSite(request); err != nil {
+		return err
 	}
-	if strings.HasPrefix(adMarkup, `{"richmedia":`) {
-		return smtAdTypeRichmedia, nil
+
+	if err := setApp(request); err != nil {
+		return err
 	}
-	if strings.HasPrefix(adMarkup, `<?xml`) {
-		return smtAdTypeVideo, nil
+
+	return setExt(request)
+}
+
+func prepareIndividualRequest(request *openrtb2.BidRequest) error {
+	imp := &request.Imp[0]
+
+	if err := setPublisherId(request, imp); err != nil {
+		return err
 	}
-	return "", fmt.Errorf("Invalid ad markup %s", adMarkup)
+
+	return setImp(imp)
 }
 
-func assignBannerSize(banner *openrtb2.Banner) (*openrtb2.Banner, error) {
-	if banner.W != nil && banner.H != nil {
-		return banner, nil
+func setUser(request *openrtb2.BidRequest) error {
+	var err error
+
+	if request.User != nil && request.User.Ext != nil {
```

Comment:
a bit odd that we require both user and user.ext to be present 
I think those two checks can be split

Reply:
okay after looking at test I got it - the config of User is done in ext